libraries/write_chef_error_handler.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+#
+# Cookbook:: aws-parallelcluster
+#
+# Copyright:: 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+module WriteChefError
+  # this class is used to handle chef errors and write the errors into a certain file if the file does not exist yet
+  class WriteChefError < Chef::Handler
+    def report
+      extend Chef::Mixin::ShellOut
+      error_file = node['cluster']['bootstrap_error_path']
+
+      # to avoid overwriting the error message from other mechanisms, such as the deprecated BYOS handler
+      # if the error file already exists we don't take any additional action here
+      unless File.exist?(error_file)
+        message_error = 'Failed to run chef recipe.'
+        node_type_to_log = {
+          'HeadNode' => 'Please check /var/log/chef-client.log in the head node, or check the chef-client.log in CloudWatch logs.',
+          'ComputeFleet' => 'Please check the cloud-init-output.log in CloudWatch logs.',
+        }
+        message_logs_to_check = node_type_to_log[node['cluster']['node_type']]
+        message_troubleshooting_link = 'Please refer to'\
+          ' https://docs.aws.amazon.com/parallelcluster/latest/ug/troubleshooting-v3.html#troubleshooting-v3-get-logs'\
```

Comment:
We should avoid the use of static URLs in error messages because public docs content may be moved and even may not be available at all. As an alternative: let's refer to a section in a doc without URLs. In this case: `Please refer to the Troubleshooting section in AWS ParallelCluster docs for more details on ParallelCluster logs.`

I know that we refer to public docs URLs in other places (eg error messages from CLI validators), but we should avoid it.

Reply:
Thanks for pointing this out. This was a team decision during the design stage. After discussion we agreed that it's not a blocking issue for now, and we will consider it in the future.